### ISSUE TITLE
unexpected nan with jax.vjp

### ISSUE THREAD

**Issue by chrisflesher** [google/jax#12409]
### Description

I am getting unexpected nan with the following example.

```
import jax
import jax.numpy as jnp


def fun(value):
    result = jnp.power(value - jnp.floor(value), jnp.array([0]))
    return result


vjp_value, vjp_grad = jax.vjp(fun, jnp.array([1.]))
print(vjp_value, vjp_grad(jnp.ones(vjp_value.shape))[0])
```
The problem seems to be introduced in jax 0.3.17 (the result is correct in 0.3.16 or earlier).

### What jax/jaxlib version are you using?

jax 0.3.17, jaxlib 0.3.15

### Which accelerator(s) are you using?

CPU

### Additional System Info

Python 3.8, Ubuntu 20.04

**Comment by chrisflesher** [comment]
My workaround for now is to wrap `jnp.power(...)` with `jnp.nan_to_num(jnp.power(...))`, this seems to return the correct result.

**Comment by jakevdp** [comment]
I think NaN is probably the most correct possible result here. Symbolically, this is the function you're defining:

$$
f(x) = (x - \lfloor x\rfloor) ^ 0
$$

Taking the derivative, we use the chain rule and find the following:

$$
f^\prime (x) = (1 - \frac{d\lfloor x\rfloor}{dx}) \cdot 0 \cdot (x - \lfloor x\rfloor) ^ {-1}
$$

The only tricky part here is the derivative of $d\lfloor x\rfloor$; this is a step function, and similar to the reasons discussed [here](https://jax.readthedocs.io/en/latest/faq.html#why-are-gradients-zero-for-functions-based-on-sort-order), JAX evaluates the derivative as zero. So we have

$$
f^\prime (x) = 0 \cdot (x - \lfloor x\rfloor) ^ {-1}
$$

Plug in $x = 1$ and you get $f(1) = 0 / 0$, for which the only sensible floating point return value is `NaN`. If you want a better-behaved derivative, you'll have to use a better-behaved function.

Hope that helps!

**Comment by chrisflesher** [comment]
I think you have a good point. I am wondering if I can solve this problem by eliminating the need to differentiate with respect to `jnp.power`?

Here are some examples:

Original (returns nan): `result = jnp.power(value - jnp.floor(value), jnp.array([0]))`
This works how I want (returns zero): `result = jnp.power(value - jnp.floor(value), 0)`
I would think this should also work how I want (but returns nan): `result = jnp.power(value - jnp.floor(value), onp.array([0]))`

Why are the last two examples returning different results?

**Comment by chrisflesher** [comment]
Hmm... I found this returns zero as well:
`result = jnp.power(value - jnp.array(value, int), onp.array([0], dtype=onp.int32))`

**Comment by jakevdp** [comment]
Yeah, that's a bit strange. I think the difference comes from the implementation of `jnp.power`, which lowers to `lax.integer_pow` if the exponent is a concrete integer, and to `lax.pow` otherwise: https://github.com/google/jax/blob/aa2f898c7960c8d8cf4c297c890a08e50b101d25/jax/_src/numpy/ufuncs.py#L310-L325

These have different gradient behavior, because in the integer case exponent of zero is treated as a special case (because there's no need to have a smooth implementation with changing exponent):
```python
import jax
import jax.numpy as jnp
from jax import lax

print(jax.grad(lambda x: lax.pow(x - lax.floor(x), 0.0))(1.0))
# nan
print(jax.grad(lambda x: lax.integer_pow(x - lax.floor(x), 0))(1.0))
# 0.0
```
If you're really concerned with raising something to a static zero power, then you could use `lax.integer_pow` directly. However if you have a real-valued exponent that happens to be zero at times, then I think NaN is the right answer.

**Comment by jakevdp** [comment]
If you want to special-case an exponent of zero in your implementation, you could do something like this:
```python
def f(x, y):
  return jnp.power(x - jnp.floor(x), y)

def f_safe(x, y):
  x_safe = jnp.where(y == 0, 1.0, x - jnp.floor(x))
  return jnp.power(x_safe, y)

print(f(1.0, 0.0))
# 1.0
print(jax.grad(f)(1.0, 0.0))
# nan
print(jax.grad(f)(1.0, 0.001))
# inf

print(f_safe(1.0, 0.0))
# 1.0
print(jax.grad(f_safe)(1.0, 0.0))
# 0.0
print(jax.grad(f_safe)(1.0, 0.001))
# inf
```

**Comment by chrisflesher** [comment]
Thanks for the advice, I'll use your suggestion.

Would it be worth considering adding `lax.integer_pow` support for static numpy integer arrays to `jnp.power` similar to how `int` is handled? Maybe someone else would encounter a similar when trying to implement bi-cubic image interpolation or tri-cubic voxel grid interpolation later.

Feel free to close this issue either way, think this solution will keep my project going, thanks.

**Comment by jakevdp** [comment]
`lax.integer_pow` is structured in a way that requires the exponent to be a static integer parameter, not an array. So if the exponent is an arbitrary numpy array (even a static one), using `lax.integer_pow` would require looping over the inputs in Python. Really this all comes down to the fact that `x^y` for `(x, y) -> (0, 0)` has a poorly-defined gradient from a mathematical perspective.

**Comment by jakevdp** [comment]
I think I'm going to close this because it doesn't look like there are any action items – please let me know if you have remaining questions!

**Comment by chrisflesher** [comment]
Thank you for your help @jakevdp!
